File: app/src/main/java/com/example/elsafa/ui/shop/OffersPresenter.java
package com.example.elsafa.ui.shop;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import Model.Offer.Offer;

public class OffersPresenter {

    OnGetOffersListener listener;

    public OffersPresenter(OnGetOffersListener listener) {
        this.listener = listener;
    }

    private void getOffersFromFirebase() {
        FirebaseFirestore fStore = FirebaseFirestore.getInstance();
        fStore.collection("Offers")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            for (DocumentSnapshot ds : task.getResult().getDocuments()) {
                                listener.onSuccess(ds.toObject(Offer.class));
                            }
                        } else {
                            listener.onFailed(task.getException());
                        }
                    }
                });
    }

    public void getOffers() {
        getOffersFromFirebase();
    }
}

File: app/src/main/java/com/whiteside/cafe/ui/shop/OffersPresenter.kt
package com.whiteside.cafe.ui.shop

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.whiteside.cafe.model.Offer

class OffersPresenter(var listener: OnGetOffersListener) {
    private fun getOffersFromFirebase() {
        val fStore = FirebaseFirestore.getInstance()
        fStore.collection("Offers")
            .get()
            .addOnSuccessListener {
                for (ds in it.documents) {
                    val offer = ds.toObject(Offer::class.java)!!
                    offer.offerId = (ds.id)
                    if (offer.endTime.seconds - Timestamp.now().seconds > 0) listener.onGetOfferSuccess(
                        offer
                    )
                }
            }
            .addOnFailureListener {
                listener.onGetOfferFailed(it)
            }
    }

    fun getOffers() {
        getOffersFromFirebase()
    }
}